internal/controller/status.go

```diff
@@ -5,25 +5,60 @@ import (
 	v1 "k8s.io/api/core/v1"
 )
 
-func getCounts(statuses []v1.ContainerStatus) amaltheadevv1alpha1.ContainerCounts {
-	total := 0
-	ready := 0
-	for _, status := range statuses {
-		total += 1
-		terminatedSuccessfully := status.State.Terminated != nil && status.State.Terminated.ExitCode == 0
-		runningSuccessfully := status.State.Running != nil && status.Started != nil && *status.Started
-		containerReady := status.Ready && (terminatedSuccessfully || runningSuccessfully)
-		if containerReady {
-			ready += 1
+// countainerCounts provides from the total and completed/fully running containers in a pod.
+// The output is a tuple with the init container counts followed by the regular container counts.
+func containerCounts(pod *v1.Pod) (amaltheadevv1alpha1.ContainerCounts, amaltheadevv1alpha1.ContainerCounts) {
+	initCounts := amaltheadevv1alpha1.ContainerCounts{}
+	counts := amaltheadevv1alpha1.ContainerCounts{}
+	for _, container := range pod.Status.InitContainerStatuses {
+		containerCompleted := container.State.Terminated != nil && container.State.Terminated.ExitCode == 0 && container.State.Terminated.Reason == "Completed"
+		initCounts.Total += 1
+		if containerCompleted {
+			initCounts.Ready += 1
 		}
 	}
-	return amaltheadevv1alpha1.ContainerCounts{Ready: ready, Total: total}
+	for _, container := range pod.Status.ContainerStatuses {
+		containerRunning := container.Ready && container.State.Running != nil
+		containerCompleted := container.State.Terminated != nil && container.State.Terminated.ExitCode == 0 && container.State.Terminated.Reason == "Completed"
+		counts.Total += 1
+		if containerRunning || containerCompleted {
+			counts.Ready += 1
+		}
+	}
+	return initCounts, counts
 }
 
-func isIdle() bool {
+func podIsReady(pod *v1.Pod) bool {
+	if pod == nil || pod.GetDeletionTimestamp() != nil {
+		// A missing pod or a pod being deleted is not considered ready
+		return false
+	}
+	phaseOk := pod.Status.Phase == v1.PodSucceeded || pod.Status.Phase == v1.PodRunning
+	initCounts, counts := containerCounts(pod)
+	return initCounts.Ok() && counts.Ok() && phaseOk
+}
+
+func podIsFailed(pod *v1.Pod) bool {
+	if pod == nil || pod.GetDeletionTimestamp() != nil {
+		// A missing pod or a pod being deleted is not considered failed
+		return false
+	}
+	restartThreshold := int32(3)
```

Comment:
Shouldn't we have that as a constant ?
Is it something that we want to be configurable ? If yes, at which level ? Controller or CRD ?

Reply:
Yes I will move that as a constant. 

But sometime in the future we should have that as part of the CR. Because unfortunately k8s does not natively support such a concept. I.e. you can set your pod to either always or never restart when it fails. Sure it has exponential (or some kind of a backoff) but still once a session is failing it will keep restarting forever.